File: app/src/main/java/com/example/dropex/fragments/HomeFragment.java
package com.example.dropex.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.GridView;

import com.example.dropex.R;
import com.example.dropex.adpaters.HomeGVAdapter;
import com.example.dropex.models.ProductModel;

import java.util.ArrayList;

public class HomeFragment extends Fragment {

    GridView homeGVAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        homeGVAdapter = view.findViewById(R.id.homeGV);
        ArrayList<ProductModel> productModelArrayList = new ArrayList<>();

        productModelArrayList.add(new ProductModel("DSA", "", 213, R.drawable.sample));
        productModelArrayList.add(new ProductModel("JAVA", "", 4242, R.drawable.sample));
        productModelArrayList.add(new ProductModel("C++", "", 3454, R.drawable.sample));
        productModelArrayList.add(new ProductModel("Python", "", 434, R.drawable.sample));
        productModelArrayList.add(new ProductModel("Javascript", "", 537, R.drawable.sample));
        productModelArrayList.add(new ProductModel("DSA", "", 686, R.drawable.sample));

        HomeGVAdapter adapter = new HomeGVAdapter(view.getContext(), productModelArrayList);
        homeGVAdapter.setAdapter(adapter);
    }
}

File: app/src/main/java/com/example/dropex/fragments/HomeFragment.kt
package com.example.dropex.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.GridView
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.example.dropex.R
import com.example.dropex.adpaters.HomeGVAdapter
import com.example.dropex.database.ProductsRepository
import com.example.dropex.models.ProductModel
import kotlinx.coroutines.launch
import java.util.ArrayList

class HomeFragment : Fragment() {
    private lateinit var homeGVAdapter: GridView
    private lateinit var loadingBar: ProgressBar
    private lateinit var searchET: EditText
    private lateinit var searchBtn: CardView
    private var productsRepository = ProductsRepository()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_home, container, false)
    }

    private fun initSearch() {
        searchBtn.setOnClickListener {
            val searchKey = searchET.text.toString()
            viewLifecycleOwner.lifecycleScope.launch {
                val productsMap = productsRepository.getAllProducts()
                val productIdsInitial = productsMap.keys.toList()
                val productModelArrayListInitial = productsMap.values.toList()
                val productIds = ArrayList<String>()
                val productModelArrayList = ArrayList<ProductModel?>()
                for (i in 0 until productsMap.size) {
                    if (productModelArrayListInitial[i]!!.name.contains(searchKey, true)) {
                        productIds.add(productIdsInitial[i])
                        productModelArrayList.add(productModelArrayListInitial[i])
                    }
                }
                val adapter = HomeGVAdapter(requireContext(), productIds, productModelArrayList)
                homeGVAdapter.adapter = adapter
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        homeGVAdapter = view.findViewById(R.id.homeGV)
        loadingBar = view.findViewById(R.id.progressBar)
        searchET = view.findViewById(R.id.searchET)
        searchBtn = view.findViewById(R.id.searchBtn)

        initSearch()

        viewLifecycleOwner.lifecycleScope.launch {
            val productsMap: Map<String, ProductModel?> = productsRepository.getAllProducts()
            val productIds = productsMap.keys.toList()
            val productModelArrayList = productsMap.values.toList()
            val adapter = HomeGVAdapter(view.context, productIds, productModelArrayList)
            homeGVAdapter.adapter = adapter
            loadingBar.visibility = View.GONE
            homeGVAdapter.visibility = View.VISIBLE
        }
    }
}